### PR TITLE
Make norm a function, add special cases, and add tests

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
Updates #38 #138

File: mat64/dense_test.go [hunk 1]
```diff
@@ -227,7 +227,9 @@ func (s *S) TestSetRowColumn(c *check.C) {
 			t.Clone(a)
 			a.SetRow(ri, make([]float64, a.mat.Cols))
 			t.Sub(t, a)
-			c.Check(t.Norm(0), check.Equals, floats.Norm(row, 2))
+			if math.Abs(Norm(t, 2)-floats.Norm(row, 2)) > 1e-14 {
+				c.Errorf("Row norm mismatch")
```
Comment (kortschak) [inline mat64/dense_test.go]: Here and below output the row/col given to the floats.Norm and the norms returned by each function.

File: mat64/matrix.go [hunk 2]
```diff
@@ -781,6 +767,87 @@ func MaybeFloat(fn func() float64) (f float64, err error) {
 	return fn(), nil
 }
 
+// Norm returns the specified (induced) norm of the matrix a. See
+// https://en.wikipedia.org/wiki/Matrix_norm for the definition of an induced norm.
+//
+// Valid norms are:
+//    1 - The maximum absolute column sum
+//    2 - Frobenius norm, the square root of the sum of the squares of the elements.
+//  Inf - The maximum row sum.
```
Comment (vladimir-ch) [inline mat64/matrix.go]: Add `absolute` to match two lines above.

File: mat64/matrix.go [hunk 3]
```diff
@@ -781,6 +767,87 @@ func MaybeFloat(fn func() float64) (f float64, err error) {
...
+// Norm will panic with ErrNormOrder if an illegal norm order is specified.
+func Norm(a Matrix, norm float64) float64 {
+	var n lapack.MatrixNorm
```
Comment (kortschak) [inline mat64/matrix.go]: Factor this block L779-L802 into a prep function that is called immediately prior to each lapack64 call. At the moment, there's a lot of work being done that might not be used.

File: mat64/matrix.go [hunk 4]
```diff
@@ -781,6 +767,87 @@ func MaybeFloat(fn func() float64) (f float64, err error) {
...
+		work = make([]float64, r)
+		if norm == 1 {
+			n = lapack.MaxRowSum
+		} else if norm == math.Inf(1) {
+			n = lapack.MaxColumnSum
+		}
+	} else {
+		work = make([]float64, c)
+	}
+	if rma, ok := aMat.(RawMatrixer); ok {
```
Comment (kortschak) [inline mat64/matrix.go]: Make these a type switch.

**Comment by btracey** [comment]
Actually, I believe this fixes #38 . Are there any I have missed?

File: mat64/dense_test.go (hunk 1 above)
Comment (btracey) [inline mat64/dense_test.go]: Done.

**Comment by kortschak** [comment]
I think you are right.

File: mat64/matrix.go (hunk 2 above)
Comment (btracey) [inline mat64/matrix.go]: Done.

File: mat64/matrix.go (hunk 3 above)
Comment (btracey) [inline mat64/matrix.go]: Refactored slightly differently.

File: mat64/matrix.go (hunk 4 above)
Comment (btracey) [inline mat64/matrix.go]: Done.

**Comment by btracey** [comment]
PTAL

File: mat64/matrix.go (hunk 3 above)
Comment (kortschak) [inline mat64/matrix.go]: Try this:

```
func Norm(a Matrix, norm float64) float64 {
	aMat, aTrans := untranspose(a)
	switch aMat := aMat.(type) {
	case RawMatrixer:
		n, work := normWork(a, norm, aTrans)
		return lapack64.Lange(n, rma.RawMatrix(), work)
	case RawTriangular:
		n, work := normWork(a, norm, aTrans)
		return lapack64.Lantr(n, rma.RawTriangular(), work)
	case RawSymmetricer:
		n, work := normWork(a, norm, aTrans)
		return lapack64.Lansy(n, rma.RawSymmetric(), work)
	}

	switch norm {
	default:
		panic("unreachable")
	case 1:
		var max float64
		for j := 0; j < c; j++ {
			var sum float64
			for i := 0; i < r; i++ {
				sum += math.Abs(a.At(i, j))
			}
			if sum > max {
				max = sum
			}
		}
		return max
	case 2:
		var sum float64
		for i := 0; i < r; i++ {
			for j := 0; j < c; j++ {
				v := a.At(i, j)
				sum += v * v
			}
		}
		return math.Sqrt(sum)
	case math.Inf(1):
		var max float64
		for i := 0; i < r; i++ {
			var sum float64
			for j := 0; j < c; j++ {
				sum += math.Abs(a.At(i, j))
			}
			if sum > max {
				max = sum
			}
		}
		return max
	}
}

func normWork(a Matrix, norm float64, trans bool) (n lapack.MatrixNorm, work []float64) {
	switch norm {
	case 1:
		n = lapack.MaxColumnSum
	case 2:
		n = lapack.NormFrob
	case math.Inf(1):
		n = lapack.MaxRowSum
	default:
		panic(ErrNormOrder)
	}
	r, c := a.Dims()
	if trans {
		work = make([]float64, r)
		if norm == 1 {
			n = lapack.MaxRowSum
		} else if norm == math.Inf(1) {
			n = lapack.MaxColumnSum
		}
	} else {
		work = make([]float64, c)
	}
	return n, work
}
```

Comment (btracey) [inline mat64/matrix.go]: The problem with that relative to my current code is that the work slice only needs to be allocated in certain scenarios 1-norm for dense, but 1 or inf norm for symmetric and triangular

Comment (kortschak) [inline mat64/matrix.go]: That can be added - the code above is an algebraic refactor of what you had.

Comment (btracey) [inline mat64/matrix.go]: I see. I’ll combine our approaches.

Comment (kortschak) [inline mat64/matrix.go]:
```
func normWork(a Matrix, norm float64, trans, full bool) (n lapack.MatrixNorm, work []float64) {
	switch norm {
	case 1:
		n = lapack.MaxColumnSum
	case 2:
		n = lapack.NormFrob
	case math.Inf(1):
		n = lapack.MaxRowSum
	default:
		panic(ErrNormOrder)
	}
	r, c := a.Dims()
	if trans {
		if norm == 1 {
			n = lapack.MaxRowSum
			if !full {
				work = make([]float64, r)
			}
		} else if norm == math.Inf(1) {
			n = lapack.MaxColumnSum
			work = make([]float64, r)
		}
	} else if !full || norm == 1 {
		work = make([]float64, c)
	}
	return n, work
}
```

with full true when passing a Dense.

**Comment by btracey** [comment]
How's that?

File: mat64/matrix.go [hunk 5]
```diff
@@ -781,6 +767,103 @@ func MaybeFloat(fn func() float64) (f float64, err error) {
...
+		}
+		return lapack64.Lantr(n, rm, work)
+	case RawSymmetricer:
+		rm := rma.RawSymmetric()
+		n := normLapack(norm, aTrans)
+		if n == lapack.MaxRowSum || n == lapack.MaxColumnSum {
+			work = make([]float64, rm.N)
+		}
+		return lapack64.Lansy(n, rm, work)
+	default:
```
Comment (kortschak) [inline mat64/matrix.go]: Make the default its own switch to reduce indentation.

File: mat64/matrix.go (hunk 3 above)
Comment (btracey) [inline mat64/matrix.go]: I feel like the code I just pushed up is cleaner.

**Comment by kortschak** [comment]
I prefer mine. There's a lot of almost repetition here.

File: mat64/matrix.go (hunk 5 above)
Comment (btracey) [inline mat64/matrix.go]: Done.

**Comment by kortschak** [comment]
OK LGTM

**Comment by btracey** [comment]
Whereas mine has a clear separation between finding the norm and allocating the work. (normLapack does one thing, and it does it the same for all of the types)

**Comment by btracey** [comment]
Alright. Thanks for the suggested improvements.

**Pull request merged**